README.md
# dev_menu

DevMenu - flutter plugin.

Package made to make team members (Dev,QA,PM) life easier while in development. Easily find out device/app info, turn certain features on and off, and preview your WIP Widgets in `Test Playground`.

![Dev_menu gif](https://media.giphy.com/media/dXXS7VQrj6d6AcGU9p/giphy.gif)

# Installation

Add `dev_menu: 1.0.3` as a [dependency in your pubspec.yaml](https://flutter.dev/docs/development/packages-and-plugins/using-packages) file.

## Information

`packageName` (Mandatory) is package name (Android: applicationId). For example: _"com.example.dev_menu_example"_

_Note_: This is not needed for iOS, since there is no way of using this _String_ to get the application data.

`flags` is list of maps, contaning _title_, and _description_. They are used to turn certain features/test on and off. Working with the help of _Shared preferences_.

Example:

```
[
  {
    'title': 'Is Environment button enabled?',
    'description': 'Determine if Environment button should work.',
  },
]
```

Using `flags` in your app:

```
bool isButtonEnabled = await DevMenuHelper()
  .getSharedPreferenceBool('Is Environment button enabled?'); -> matches "title" from above.
print('Is button enabled? ' + isButtonEnabled);
```

Or use `bool` to show Widget with combination of `FutureBuilder`

```
return FutureBuilder<bool>(
  future: DevMenuHelper()
    .getSharedPreferenceBool('Should show recent transactions?'),
  builder: (context, snapshot) {
    if (snapshot.connectionState == ConnectionState.done &&
    snapshot.hasData) {
      bool shouldShowTransactions = snapshot.data;
      return shouldShowTransactions ? TransactionsWidget() : Container();
    } else {
      return Container();
    }
  },
);
```

`testWidgets` is list of widgets provided for Test playground screen of plugin. Usually widgets which are not yet connected with the Apps flow, or work in progress stuff.

Example:

```
[
  {
    'name': 'Alert dialog',
    'widget': AlertDialog(
      title: Text('Title'),
      content: Text('Content'),
      actions: <Widget>[
        FlatButton(
          onPressed: () {},
          child: Center(
            child: Text('Button 1'),
          ),
        ),
        FlatButton(
          onPressed: () {},
          child: Center(
            child: Text('Button 2'),
          ),
        )
      ],
    ),
  },
  {
    'name': 'Sized box',
    'widget': SizedBox(
      width: 300,
      height: 400,
      child: Container(
        color: Colors.red,
        child: Text('I am text in sized box'),
      ),
   )
  }
]
```

_Application info_:

| Name             | Label       | iOS           | Android                                                                                                    |
| ---------------- | ----------- | ------------- | ---------------------------------------------------------------------------------------------------------- |
| Application name | appName     | not available | [android app name](https://developer.android.com/guide/topics/manifest/manifest-intro#package-name)        |
| Package name     | packageName | not available | [android package name](https://developer.android.com/reference/android/content/pm/PackageInfo#packageName) |
| Version          | versionName | not available | [android version](https://developer.android.com/reference/android/content/pm/PackageInfo#versionName)      |

Here you can also see any custom application info, provided by `customAppInfo`

`customAppInfo` is list of custom informations regarding the app. Contains _title_, initially _selected_ option, and list of all _options_. Using _Shared preferences_, selected will be read on the first run, afterwards selected is read from the phone memory.

Example:

```
[
  {
    'title': 'Environment',
    'selected': 'QA',
    'options': ['QA', 'DEV', 'PROD', 'STAGING']
  },
]
```

_Device info_:

Android:

| Name                  | description                                                                                                                                                       |
| --------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| version.securityPatch | [The user-visible security patch level](https://developer.android.com/reference/android/os/Build.VERSION#SECURITY_PATCH)                                          |
| version.sdkInt        | [The user-visible SDK version of the framework](https://developer.android.com/reference/android/os/Build.VERSION#PREVIEW_SDK_INT)                                 |
| version.release       | [The user-visible version string](https://developer.android.com/reference/android/os/Build.VERSION#RELEASE)                                                       |
| version.codename      | [The current development codename, or the string "REL" if this is a release build](https://developer.android.com/reference/android/os/Build.VERSION#CODENAME)     |
| version.baseOs        | [The base OS build the product is based on](https://developer.android.com/reference/android/os/Build.VERSION#BASE_OS)                                             |
| device                | [The name of the industrial design](https://developer.android.com/reference/android/os/Build#DEVICE)                                                              |
| manufacturer          | [The manufacturer of the product/hardware](https://developer.android.com/reference/android/os/Build#MANUFACTURER)                                                 |
| model                 | [The end-user-visible name for the end product](https://developer.android.com/reference/android/os/Build#MODEL)                                                   |
| product               | [The name of the overall product](https://developer.android.com/reference/android/os/Build#PRODUCT)                                                               |
| androidId             | [The Android hardware device ID that is unique between the device + user and app signing](<https://developer.android.com/reference/android/os/Build#getSerial()>) |

iOS:

| Name                | description                                                                                                                              |
| ------------------- | ---------------------------------------------------------------------------------------------------------------------------------------- |
| name                | [Device name](https://developer.apple.com/documentation/uikit/uidevice/1620015-name)                                                     |
| systemName          | [The name of the current operating system](https://developer.apple.com/documentation/uikit/uidevice/1620054-systemname)                  |
| systemVersion       | [The current operating system version](https://developer.apple.com/documentation/uikit/uidevice/1620043-systemversion)                   |
| model               | [Device model](https://developer.apple.com/documentation/uikit/uidevice/1620044-model)                                                   |
| identifierForVendor | [Unique UUID value identifying the current device](https://developer.apple.com/documentation/uikit/uidevice/1620059-identifierforvendor) |
| utsname.sysname     | [Operating system name](https://pubs.opengroup.org/onlinepubs/7908799/xsh/sysutsname.h.html)                                             |
| utsname.nodename    | [Network node name](https://pubs.opengroup.org/onlinepubs/7908799/xsh/sysutsname.h.html)                                                 |
| utsname.release     | [Release level](https://pubs.opengroup.org/onlinepubs/7908799/xsh/sysutsname.h.html)                                                     |
| utsname.version     | [Version level](https://pubs.opengroup.org/onlinepubs/7908799/xsh/sysutsname.h.html)                                                     |
| utsname.machine     | [Hardware type](https://pubs.opengroup.org/onlinepubs/7908799/xsh/sysutsname.h.html)                                                     |


CHANGELOG.md
## 1.0.4

- Add more examples for Flags

## 1.0.3

- Fixed text errors

## 1.0.2

- Updated description


lib/dev_menu.dart
import 'package:dev_menu/widgets/info_list.dart';
import 'package:dev_menu/widgets/playground.dart';
import 'package:flutter/material.dart';

import 'package:dev_menu/helpers/shared_preferences_helper.dart';
import 'package:dev_menu/widgets/flags_list.dart';
import 'package:dev_menu/helpers/device_info_helper.dart';
import 'package:flutter/services.dart';

class DevMenuHelper with SharedPreferencesHelper {
  @override
  Future<bool> setSharedPreferenceBool(String title, bool value) {
    return super.setSharedPreferenceBool(title, value);
  }

  @override
  Future<bool> getSharedPreferenceBool(String title) {
    return super.getSharedPreferenceBool(title);
  }

  @override
  Future<bool> setSharedPreferenceString(String title, String value) {
    return super.setSharedPreferenceString(title, value);
  }

  @override
  Future<String> getSharedPreferenceString(String title) {
    return super.getSharedPreferenceString(title);
  }
}

class DevMenu extends StatefulWidget {
  /// DevMenu plugin
  ///
  /// [packageName] is package name (Android: applicationId).
  /// For example: "com.example.dev_menu_example"
  /// Note: This is not needed for iOS, since there is no way of using this
  /// String to get the application data.
  ///
  ///
  /// [flags] is [List] of [Map]s, contaning title, and description.
  /// They are used to turn certain features/test on and off. Working with
  /// the help of Shared prefferences.
  /// Example:
  /// [
  ///   {
  ///     'title': 'Is Environment button enabled?',
  ///     'description': 'Determine if Environment button should work.',
  ///   },
  /// ]
  ///
  ///
  /// [testWidgets] is list of [Widget]s provided for Test playground screen
  /// of plugin. Usually widgets which are not yet connected
  /// with the Apps flow, or work in progress stuff.
  /// Example:
  /// [
  ///   {
  ///     'name': 'Alert dialog',
  ///     'widget': AlertDialog(
  ///       title: Text('Title'),
  ///       content: Text('Content'),
  ///       actions: <Widget>[
  ///         FlatButton(
  ///           onPressed: () {},
  ///           child: Center(
  ///             child: Text('Button 1'),
  ///           ),
  ///         ),
  ///         FlatButton(
  ///           onPressed: () {},
  ///           child: Center(
  ///             child: Text('Button 2'),
  ///           ),
  ///         )
  ///       ],
  ///     ),
  ///   },
  ///   {
  ///     'name': 'Sized box',
  ///     'widget': SizedBox(
  ///       width: 300,
  ///       height: 400,
  ///       child: Container(
  ///         color: Colors.red,
  ///         child: Text('I am text in sized box'),
  ///       ),
  ///     )
  ///   }
  /// ]
  ///
  ///
  /// [customAppInfo] is list of custom informations regarding the app.
  /// Contains title, initially selected option, and list of all options.
  /// Using Shared preferences, selected will be read on the first run,
  /// afterwards selected is read from the phone memory.
  /// Example:
  /// [
  ///   {
  ///     'title': 'Environment',
  ///     'selected': 'QA',
  ///     'options': ['QA', 'DEV', 'PROD', 'STAGING']
  ///   },
  /// ],
  ///
  ///

  final String packageName;
  final List<Map<String, dynamic>> flags;
  final List<Map<String, dynamic>> testWidgets;
  final List<Map<String, dynamic>> customAppInfo;

  const DevMenu({
    Key key,
    this.flags,
    this.packageName,
    this.testWidgets,
    this.customAppInfo,
  }) : super(key: key);

  @override
  _DevMenuState createState() => _DevMenuState();
}

class _DevMenuState extends State<DevMenu> with DeviceInfoHelper {
  Map<String, dynamic> _deviceData = <String, dynamic>{};
  Map _appData;

  static MethodChannel packageInfoPlatform =
      const MethodChannel('devmenu.flutter.io/packageInfo');

  @override
  void initState() {
    super.initState();

    _getDeviceInfo();
    _getPackageInfo();
  }

  Future<void> _getPackageInfo() async {
    if (widget.packageName != null) {
      await packageInfoPlatform.invokeMethod('getPackageInfo', {
        "packageName": widget.packageName,
      }).then((result) {
        setState(() {
          _appData = result;
        });
      }).catchError((e) {
        print(e);
      });
    }
  }

  Future<void> _getDeviceInfo() async {
    var data = await initPlatformState();

    setState(() {
      _deviceData = data;
    });
  }

  @override
  Widget build(BuildContext context) {
    List<Map<String, dynamic>> screens = [
      {
        'name': 'Flags',
        'widget': FlagsList(list: widget.flags),
        'description':
            'Switches to turn certain features or test options on and off.',
      },
      {
        'name': 'Application info',
        'widget': InfoList(
          data: _appData,
          title: 'Application info',
          customAppInfo: widget.customAppInfo,
        ),
        'description':
            'Find out (and add your own, for example the environment) some useful info about the app like app version, build number.',
      },
      {
        'name': 'Device info',
        'widget': InfoList(
          data: _deviceData,
          title: 'Device info',
        ),
        'description':
            'Find out some useful info about the device like manufacturer, OS version, phone dimensions and many more.',
      },
      {
        'name': 'Test playground',
        'widget': Playground(widgets: widget.testWidgets),
        'description':
            'Add the WIP components of your app or the ones that are not connected to flow so other members of the team (QAs can test the UI for example) can see it.',
      }
    ];

    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: Text('Developer menu'),
      ),
      body: ListView(
        physics: NeverScrollableScrollPhysics(),
        children: screens
            .map(
              (item) => ListTile(
                isThreeLine: item['name'] != 'Flags',
                subtitle: item['description'] != null
                    ? Text(item['description'])
                    : Container(),
                title: Text(item['name']),
                onTap: () {
                  Navigator.push(
                    context,
                    MaterialPageRoute(
                      builder: (context) => item['widget'],
                    ),
                  );
                },
              ),
            )
            .toList(),
      ),
    );
  }
}
